### PR TITLE
Implement parts of W3DView

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/ingameui.h [hunk 1]
```diff
@@ -227,6 +227,7 @@ class InGameUI : public SubsystemInterface, public SnapShot
     bool Get_Unk7() const { return m_unk7; }
     void Set_Unk6(bool b) { m_unk6 = b; }
     void Clear_Attack_Move_To_Mode() { m_attackMoveToMode = false; }
+    bool Is_In_Force_To_Attack_Mode() { return m_forceToAttackMode; }
```
Comment (xezon) [inline src/game/client/ingameui.h:230]: Can be const

File: src/platform/w3dengine/client/w3dview.cpp [hunk 2]
```diff
@@ -13,7 +13,471 @@
...
+#include "globaldata.h"
+#include "ingameui.h"
+#include "object.h"
+#include "scriptengine.h"
+#include "selectioninfo.h"
 #include "shadermanager.h"
+#include "terrainlogic.h"
+#include "w3ddisplay.h"
+#include "w3dscene.h"
+static int g_theW3DFrameLengthInMsec = 33;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: const

File: src/platform/w3dengine/client/w3dview.cpp [hunk 3]
```diff
@@ -13,7 +13,471 @@
...
+void W3DView::Set_Camera_Transform()
+{
+    m_cameraHasMovedSinceRequest = true;
+    Matrix3D tm(true);
+    float znear = 10.0f;
+    float zfar = 1200.0f;
+
+    if (m_realZoom) {
+        if (m_FXPitch < 0.95f) {
+            zfar = zfar / m_FXPitch;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: Or /= m_FXPitch

File: src/platform/w3dengine/client/w3dview.cpp [hunk 4]
```diff
@@ -13,7 +13,471 @@
...
+        }
+    } else if ((g_theWriteableGlobalData != nullptr && g_theWriteableGlobalData->m_drawEntireTerrain) || m_FXPitch < 0.95f
+        || m_zoom > 1.05f) {
+        zfar = zfar * 10.0f;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: Or *= 10.0f

File: src/platform/w3dengine/client/w3dview.cpp [hunk 5]
```diff
@@ -13,7 +13,471 @@
...
+void W3DView::Build_Camera_Transform(Matrix3D *transform)
+{
+    Vector3 p;
+    Vector3 t;
+    float zoom = Get_Zoom();
+    float angle = Get_Angle();
+    float pitch = Get_Pitch();
+    Coord3D pos = Get_Position();
+    pos.x = pos.x + m_shakeOffset.x;
+    pos.y = pos.y + m_shakeOffset.y;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: Or +=
2x

File: src/platform/w3dengine/client/w3dview.cpp [hunk 6]
```diff
@@ -13,7 +13,471 @@
...
+
+        if (real_zoom > 1.0f) {
+            real_zoom = 1.0f;
+        }
+
+        if (real_zoom < 0.5f) {
+            real_zoom = 0.5f;
+        }
+
+        m_FOV = 0.87266463f * real_zoom * real_zoom;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:215]: Or PI / 3.6f

File: src/platform/w3dengine/client/w3dview.cpp [hunk 7]
```diff
@@ -13,7 +13,471 @@
...
+    m2.Mul_Vector3(p);
+    m1.Mul_Vector3(p);
+
+    p *= k;
+    p.X = p.X + pos.x;
+    p.Y = p.Y + pos.y;
+    p.Z = p.Z + m_groundLevel;
+    t.X = t.X + pos.x;
+    t.Y = t.Y + pos.y;
+    t.Z = t.Z + m_groundLevel;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: Or +=
6x

File: src/platform/w3dengine/client/w3dview.cpp [hunk 8]
```diff
@@ -13,7 +13,471 @@
...
+
+    return drawable;
+}
+
+int W3DView::Iterate_Drawables_In_Region(IRegion2D *screen_region, bool (*callback)(Drawable *, void *), void *user_data)
+{
+    int count = 0;
+    Vector3 dest;
+    Vector3 ws_point;
+    bool b = false;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: is_point

File: src/platform/w3dengine/client/w3dview.cpp [hunk 9]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+        Rotate_Camera_One_Frame();
+        ret = true;
+    } else if (m_doingMoveCameraOnWaypointPath) {
+        m_previousLookAtPosition = Get_Position();
+        Move_Along_Waypoint_Path(g_theW3DFrameLengthInMsec);
+        ret = true;
+    }
+
+    if (m_doingScriptedCameraLock) {
+        return true;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:518]: Or `ret = true`

File: src/platform/w3dengine/client/w3dview.cpp [hunk 10]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+{
+    if (*angle < -31.415928f) {
+        *angle = 0.0f;
+    }
+
+    if (*angle > 31.415928f) {
+        *angle = 0.0f;
+    }
+
+    while (*angle < -3.1415927f) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:556]: Or PI

File: src/platform/w3dengine/client/w3dview.cpp [hunk 11]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+        *angle = *angle + 6.2831855f;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: Or PI * 2.f

File: src/platform/w3dengine/client/w3dview.cpp [hunk 12]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+    }
+}
+
+void Norm_Angle(float *angle)
+{
+    if (*angle < -31.415928f) {
+        *angle = 0.0f;
+    }
+
+    if (*angle > 31.415928f) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:552]: Or PI * 10.f

File: src/platform/w3dengine/client/w3dview.cpp [hunk 13]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+        *angle = 0.0f;
+    }
+
+    while (*angle < -3.1415927f) {
+        *angle = *angle + 6.2831855f;
+    }
+
+    while (*angle > 3.1415927f) {
+        *angle = *angle - 6.2831855f;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: Or -=, +=
2x

File: src/platform/w3dengine/client/w3dview.cpp [hunk 14]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+                int len = v.Length();
+
+                if (len >= 0.1f) {
+                    float angle = GameMath::Acos(v.X / len);
+
+                    if (v.Y < 0.0f) {
+                        angle = -angle;
+                    }
+
+                    angle -= 1.5707964f;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:593]: Or PI * 0.5f

File: src/platform/w3dengine/client/w3dview.cpp [hunk 15]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+                f6 = f6 + (1.0f - f6) * f7;
+            }
+
+            float f8 = 1.0f - f7;
+            float f9 = 1.0f - f8;
+            float f10 = m_mcwpInfo.camera_angle[m_mcwpInfo.cur_segment];
+            float f11 = m_mcwpInfo.camera_angle[m_mcwpInfo.cur_segment + 1];
+
+            if (f11 - f10 > GAMEMATH_PI) {
+                f10 = f10 + GAMEMATH_PI2;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: Or +=

File: src/platform/w3dengine/client/w3dview.cpp [hunk 16]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+            }
+
+            if (f11 - f10 < -GAMEMATH_PI) {
+                f10 = f10 - GAMEMATH_PI2;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: Or -=

File: src/platform/w3dengine/client/w3dview.cpp [hunk 17]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+            }
+
+            float f12 = f10 * f8 + f11 * f9;
+            Norm_Angle(&f12);
+            float f13 = f12 - m_angle;
+            Norm_Angle(&f13);
+
+            if (GameMath::Fabs(f13) > 0.3141592741012573f) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:732]: Or PI / 10.f

File: src/platform/w3dengine/client/w3dview.cpp [hunk 18]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+            Norm_Angle(&m_angle);
+            float f14 = m_mcwpInfo.time_multiplier[m_mcwpInfo.cur_segment] * f8
+                + m_mcwpInfo.time_multiplier[m_mcwpInfo.cur_segment + 1] * f9;
+            m_timeMultiplier = GameMath::Fast_To_Int_Floor(f14 + 0.5f);
+            m_groundLevel = f8 * m_mcwpInfo.ground_height[m_mcwpInfo.cur_segment]
+                + f9 * m_mcwpInfo.ground_height[m_mcwpInfo.cur_segment + 1];
+            m_cameraOffset.y =
+                -(m_cameraOffset.z / GameMath::Tan(g_theWriteableGlobalData->m_cameraPitch * 0.01745329300562541f));
+            m_cameraOffset.x =
+                -(GameMath::Tan(g_theWriteableGlobalData->m_cameraYaw * 0.01745329300562541f) * m_cameraOffset.y);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:746]: Or PI / 180.f
4x

File: src/platform/w3dengine/client/w3dview.cpp [hunk 19]
```diff
@@ -27,9 +491,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+        m_angle = m_mcwpInfo.camera_angle[m_mcwpInfo.num_waypoints];
+        m_groundLevel = m_mcwpInfo.ground_height[m_mcwpInfo.num_waypoints];
+        m_cameraOffset.y =
+            -(m_cameraOffset.z / GameMath::Tan(g_theWriteableGlobalData->m_cameraPitch * 0.01745329300562541f));
+        m_cameraOffset.x = -(GameMath::Tan(g_theWriteableGlobalData->m_cameraYaw * 0.01745329300562541f) * m_cameraOffset.y);
+
+        Coord3D pos;
+        pos.x = m_mcwpInfo.waypoints[m_mcwpInfo.num_waypoints].x;
+        pos.y = m_mcwpInfo.waypoints[m_mcwpInfo.num_waypoints].y;
+        pos.z = 0.0;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: f

File: src/platform/w3dengine/client/w3dview.cpp [hunk 20]
```diff
@@ -13,7 +13,471 @@
...
+        p.Y = (p.Y - t.Y) / m_FXPitch + t.Y;
+    } else {
+        float fxpitch = p.Z - t.Z;
+        fxpitch *= m_FXPitch;
+        t.Z = p.Z - fxpitch;
+    }
+
+    transform->Make_Identity();
+    transform->Look_At(p, t, 0.0f);
+    g_theCameraShakerSystem.Timestep(0.033333335f);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:271]: Or 1.f / 30.f

File: src/platform/w3dengine/client/w3dview.cpp [hunk 21]
```diff
@@ -13,7 +13,471 @@
...
+    Vector2 min;
+    Vector2 max;
+    float f1 = width;
+    float width_to_height = f1 / Get_Height();
+    m_3DCamera->Set_Aspect_Ratio(width_to_height);
+    m_3DCamera->Get_Viewport(min, max);
+    float f2 = (float)(width + m_originX);
+    max.X = f2 / g_theDisplay->Get_Width();
+    m_3DCamera->Set_Viewport(min, max);
+    float f3 = f1 / g_theDisplay->Get_Width() * 0.87266463f;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp:452]: Or PI / 3.6f

**Comment by xezon** [comment]
Camera in Shell Map moves too fast and moves to invalid places

shot_20231229_191616_2

File: src/platform/w3dengine/client/w3dview.cpp (hunk 2 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 3 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 4 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 5 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 7 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 8 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 13 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 15 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 16 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

File: src/platform/w3dengine/client/w3dview.cpp (hunk 19 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

**Comment by jonwil** [comment]
I sorted all the comments.
The camera problem on the shell map is caused by W3DView::Move_Along_Waypoint_Path. Been staring at that math for a while with no luck spotting what's wrong and have asked Toms for help.

**Comment by jonwil** [comment]
Camera issues on shell map seem to be fixed.

File: src/platform/w3dengine/client/w3dview.cpp [hunk 22]
```diff
@@ -27,9 +492,300 @@ void W3DView::Set_View_Filter_Pos(const Coord3D *pos)
...
+    if (*angle < -31.415928f) {
+        *angle = 0.0f;
+    }
+
+    if (*angle > 31.415928f) {
+        *angle = 0.0f;
+    }
+
+    while (*angle < -3.1415927f) {
+        *angle -= 6.2831855f;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dview.cpp]: +=, otherwise this loops infinitely.

Comment (jonwil) [reply src/platform/w3dengine/client/w3dview.cpp]: fixed.

**Review by xezon — COMMENTED** [review]
In the GLA01 campaign intro, there is a black frame shortly before the jeep crashes into the tower. This does not happen with Thyme prior this change.

Edit: Appears to be introduced by **W3DView::Rotate_Camera_One_Frame**

https://github.com/TheAssemblyArmada/Thyme/assets/4720891/02964512-3ed5-424f-98ca-fe5ecc271d43

**Comment by jonwil** [comment]
Issue with GLA01 should be fixed.

**Review by xezon — APPROVED** [review]
Shell Map looks as usual. Intro sequences in USA01, GLA01 and China01 look normal. Camera controls act normal. No crashes.

**Pull request merged**